### DISCUSSION TITLE
How to import the HeteroData from .csv file containing triplet?

### DISCUSSION THREAD

**Discussion by songsong0425** [pyg-team/pytorch_geometric#7349] [Q&A] [answered]
Dear PyG community,

Greetings, I have questions about the knowledge graph embedding(KGE) using pytorch geometric.

When I have a knowledge graph with a .csv file containing triplet as below, how can I import it as HeteroData format? 
I want to use KGE in torch_geometric using my dataset (reference code: https://github.com/pyg-team/pytorch_geometric/blob/master/examples/kge_fb15k_237.py).
For the exercise, I attached the example knowledge graph file. If anyone needs a dataset for reproducing, please use it.

[kg_ex.csv](https://github.com/pyg-team/pytorch_geometric/files/11449143/kg_ex.csv)

Note that there were any initial feature vectors for entities and relations. In this case, how can features be initialized?

**Comment by wsad1** [comment, marked as answer]
Pyg doesn't have an inbuilt functions to generate `HeteroData` object from the kind of data you shared. You'll have to write your own functions which does something like
1. Read the csv as a pandas dataframe. 
2. For each unique `head_label` and `tail_label` pair construct an `edge_index`, and add it to `HeteroData` object. 

Now since you don't have node features, you could
1. Construct node features like `degree`, `centrality` etc
2. If this is an inductive task you could use [Embedding](https://pytorch.org/docs/stable/generated/torch.nn.Embedding.html)

**Reply by songsong0425** [reply]
Thank you for your reply. Then, should head- and tail-label have independent id? For example, which one is correct, Human{A:0, B:1}~Animal{C:0, D:1} or Human{A:0, B:1}~Animal{C:2, D:3}?

**Reply by rusty1s** [reply]
Each node type currently has its own set of IDs in a `HeteroData` object. Alternatively, you can cast this as a classical homogeneous graph, where you attach `node_type` and `edge_type` information as node-level and edge-level attributes (this is what most KGE approaches do, check `examples/kge_fb15k_237.py`). Hope this helps.

**Reply by songsong0425** [reply]
Always thank you for your reply. I understand your answer.
Besides, can I ask some questions? Although I checked the fb15k_237 dataset (pic 1), I couldn't find the code how to convert the HeteroData to Data. Also, should I prepare train, valid and test manually from HeteroData (or Data) before run the `examples/kge_fb15k_237.py`? 
![fb15k_237](https://github.com/pyg-team/pytorch_geometric/assets/80442548/0e188be6-a048-45f2-b1cd-c68090bcdc6c)

+) Did you mean to use `to_homogeneous` described in https://pytorch-geometric.readthedocs.io/en/latest/generated/torch_geometric.data.HeteroData.html ?

**Reply by rusty1s** [reply]
You can find the source code here (https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/datasets/freebase.py#L64), which should be pretty similar to what you want to do. I personally suggest to represent this as a normal `Data` object if you do not have any features.

**Reply by songsong0425** [reply]
Thank you for your suggestion. For days, I tried to modify the `freebase.py` for my code, but there was an unexpected error as below:
`RuntimeError: Expected tensor for argument #1 'indices' to have one of the following scalar types: Long, Int; but got torch.cuda.FloatTensor instead (while checking arguments for embedding)` .
I only changed the code associated with train/val/test splitting and edge_index & edge_type initialization.

1. For edge_index and edge_type initialization, `freebase.py` used `torch.empty`. Otherwise, I used `torch.zeros` and matched values according to the values in the dictionaries (i.e., entity and relation dictionary). Also, I used `range` for loof because `enumerate` evoked an error.

```python
import numpy as np
import pandas as pd
import torch
from torch import nn, optim
from torch_geometric.data import Data

edge_index = torch.zeros((2, len(mapped_relation_df)), dtype=torch.long)
edge_type = torch.zeros(len(mapped_relation_df), dtype=torch.long)

for i in range(len(mapped_relation_df)):
    edge_index[0,i] = mapped_relation_df.iloc[i][0]
    edge_index[1,i] = mapped_relation_df.iloc[i][2]
    edge_type[i] = mapped_relation_df.iloc[i][1]
    
data_ex = Data(edge_index=edge_index, edge_type=edge_type)
```
![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/f36221fb-95f9-4cd7-afbe-0fe3cc20058a)

2. After that, I called a train, validation, and test dataset using `RandomLinkSplit` in `torch_geometric.transforms`. Note that the original `freebase.py` called them from separate files.

```python
import torch_geometric.transforms as T

transform = T.RandomLinkSplit(num_val=0.1, num_test=0.2, is_undirected=False)
tr_d, val_d, test_d = transform(data_ex)
```
![image](https://github.com/pyg-team/pytorch_geometric/assets/80442548/f608776d-4b32-401f-ab69-d16e16ad10f9)

When I searched this error in Google, it is due to the type of tensor. But as the above code, I set the tensor type as `torch.long`.
Although I think that I almost followed the original scheme of `freebase.py` to run `kge_fb15k_237.py`, can you tell me where should I fix it?

**Reply by songsong0425** [reply]
I found that `tr_d.edge_label.dtype` is set to torch.float32. I think that it would be better to set the type of `edge_label` as torch.long (or torch.int64) automatically.

**Reply by rusty1s** [reply]
Can you clarify where exactly the error occurs?

**Reply by songsong0425** [reply]
For what it's worth, I attach the error log. In summary, `edge_label` is set to `torch.float32` after running `RandomLinkSplit` due to the default type of Python. Although I should set the type of `edge_label` like `edge_index = torch.zeros((2, len(mapped_relation_df)), dtype=torch.long)`, it would be better to set the type of `edge_label` to `torch.long` or `torch.int64` automatically or add a type setting in `freebase.py` file. 

```python
---------------------------------------------------------------------------
RuntimeError                              Traceback (most recent call last)
Cell In [38], line 2
      1 for epoch in range(1, 501):
----> 2     loss = train()
      3     print(f"Epoch: {epoch:03d}, Loss: {loss:.4f}")
      4     if epoch % 25 == 0:

Cell In [36], line 6, in train()
      4 for head_index, rel_type, tail_index in loader:
      5     optimizer.zero_grad()
----> 6     loss = model.loss(head_index, rel_type, tail_index)
      7     loss.backward()
      8     optimizer.step()

File ~/.venv/torch1.12.1-py3.8-cuda11.3/lib/python3.8/site-packages/torch_geometric/nn/kge/transe.py:90, in TransE.loss(self, head_index, rel_type, tail_index)
     83 def loss(
     84     self,
     85     head_index: Tensor,
     86     rel_type: Tensor,
     87     tail_index: Tensor,
     88 ) -> Tensor:
---> 90     pos_score = self(head_index, rel_type, tail_index)
     91     neg_score = self(*self.random_sample(head_index, rel_type, tail_index))
     93     return F.margin_ranking_loss(
     94         pos_score,
     95         neg_score,
     96         target=torch.ones_like(pos_score),
     97         margin=self.margin,
     98     )

File ~/.venv/torch1.12.1-py3.8-cuda11.3/lib/python3.8/site-packages/torch/nn/modules/module.py:1130, in Module._call_impl(self, *input, **kwargs)
   1126 # If we don't have any hooks, we want to skip the rest of the logic in
   1127 # this function, and just call forward.
   1128 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130     return forward_call(*input, **kwargs)
   1131 # Do not call functions when jit is used
   1132 full_backward_hooks, non_full_backward_hooks = [], []

File ~/.venv/torch1.12.1-py3.8-cuda11.3/lib/python3.8/site-packages/torch_geometric/nn/kge/transe.py:74, in TransE.forward(self, head_index, rel_type, tail_index)
     66 def forward(
     67     self,
     68     head_index: Tensor,
     69     rel_type: Tensor,
     70     tail_index: Tensor,
     71 ) -> Tensor:
     73     head = self.node_emb(head_index)
---> 74     rel = self.rel_emb(rel_type)
     75     tail = self.node_emb(tail_index)
     77     head = F.normalize(head, p=self.p_norm, dim=-1)

File ~/.venv/torch1.12.1-py3.8-cuda11.3/lib/python3.8/site-packages/torch/nn/modules/module.py:1130, in Module._call_impl(self, *input, **kwargs)
   1126 # If we don't have any hooks, we want to skip the rest of the logic in
   1127 # this function, and just call forward.
   1128 if not (self._backward_hooks or self._forward_hooks or self._forward_pre_hooks or _global_backward_hooks
   1129         or _global_forward_hooks or _global_forward_pre_hooks):
-> 1130     return forward_call(*input, **kwargs)
   1131 # Do not call functions when jit is used
   1132 full_backward_hooks, non_full_backward_hooks = [], []

File ~/.venv/torch1.12.1-py3.8-cuda11.3/lib/python3.8/site-packages/torch/nn/modules/sparse.py:158, in Embedding.forward(self, input)
    157 def forward(self, input: Tensor) -> Tensor:
--> 158     return F.embedding(
    159         input, self.weight, self.padding_idx, self.max_norm,
    160         self.norm_type, self.scale_grad_by_freq, self.sparse)

File ~/.venv/torch1.12.1-py3.8-cuda11.3/lib/python3.8/site-packages/torch/nn/functional.py:2199, in embedding(input, weight, padding_idx, max_norm, norm_type, scale_grad_by_freq, sparse)
   2193     # Note [embedding_renorm set_grad_enabled]
   2194     # XXX: equivalent to
   2195     # with torch.no_grad():
   2196     #   torch.embedding_renorm_
   2197     # remove once script supports set_grad_enabled
   2198     _no_grad_embedding_renorm_(weight, input, max_norm, norm_type)
-> 2199 return torch.embedding(weight, input, padding_idx, scale_grad_by_freq, sparse)

RuntimeError: Expected tensor for argument #1 'indices' to have one of the following scalar types: Long, Int; but got torch.FloatTensor instead (while checking arguments for embedding)
```

**Reply by rusty1s** [reply]
`edge_label` is of `dtype==torch.float` to align with the usage of `torch.nn.functional.binary_cross_entropy`. From the code-block you posted above, it is not totally clear to me where `edge_label` is actually used on your end.

**Reply by songsong0425** [reply]
I almost followed the https://github.com/pyg-team/pytorch_geometric/blob/master/examples/kge_fb15k_237.py code. 

In my answer written three days ago, I splitted Data into train:valid:test = 7:2:1 using `RandomLinkSplit` and the number of `edge_label` represented them (i.e., 3022954, 431850, 863700, repectively).
> ```python
> import torch_geometric.transforms as T
> 
> transform = T.RandomLinkSplit(num_val=0.1, num_test=0.2, is_undirected=False)
> tr_d, val_d, test_d = transform(data_ex)
> ```
> 
> ![image](https://user-images.githubusercontent.com/80442548/238356710-f608776d-4b32-401f-ab69-d16e16ad10f9.png)

I don't know why `RandomLinkSplit` returned not only `edge_index` and `edge_type` but also `edge_label` and `edge_label_index`.
But `edge_label` and `edge_label_index` look as splitted relations and entities, I used them as `edge_type` and `edge_index[x]` in kge_fb15k_237.py file, respectively. (Also, I referenced the discussion in #3668 )
> ![image](https://user-images.githubusercontent.com/80442548/238355959-f36221fb-95f9-4cd7-afbe-0fe3cc20058a.png)

```python
# Original kge_fb15k_237.py
loader = model.loader(
    head_index=train_data.edge_index[0], #
    rel_type=train_data.edge_type, # 
    tail_index=train_data.edge_index[1], #
    batch_size=1000,
    shuffle=True,
)
```
```python
# Customized code
loader = model.loader(
    head_index=train_data.edge_label_index[0], #
    rel_type=train_data.edge_label, #
    tail_index=train_data.edge_label_index[1], #
    batch_size=1000,
    shuffle=True,
)
```

In conclusion, I used `edge_label` in data loader for running KGE model comparison code.

**Reply by rusty1s** [reply]
Understood. The `RandomLinkSplit` is not really designed for KGE-usage. I think you should just implement your own random-link split logic in that case, e.g.
```python
perm = torch.randperm(data.edge_index.size(1))
test_idx = perm[1000:]
train_idx = perm[:1000]

model.loader(
    head_index = data.edge_index[0, train_idx],
    rel_type = data.edge_type[train_idx],
    tail_index = data.edge_index[1, train_idx],
    ...,
)

**Reply by songsong0425** [reply]
Okay. I'll try to change my code for splitting datasets with your suggestion. Thank you so much!
